### PR TITLE
[ActionSheet] Add properties to allow for theming

### LINKED ISSUE #4903: [ActionSheet] Expose properties to enable theming
Expose all properties to enable color, typography and shape theming.

### PR DESCRIPTION
**Pull request by codeman7**
This PR adds properties to the components to allow for theming.

Closes #4903

### CODE REVIEW

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m [hunk 1]
```diff
@@ -148,14 +148,13 @@ - (void)layoutSubviews {
...
-
-  _imageView.image = _itemAction.image;
+  _imageView.image = [_itemAction.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
 }
 
 - (void)setAction:(MDCActionSheetAction *)action {
   _itemAction = [action copy];
   _textLabel.text = _itemAction.title;
-  _imageView.image = _itemAction.image;
+  _imageView.image = [_itemAction.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m]: Let's allow clients choose their rendering mode. They may have a branded image that should retain its full color palette.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 2]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
...
+/**
+ The color applied to the sheet view of the action sheet controller.
+ */
 @property(nonatomic, nonnull, strong) UIColor *backgroundColor;
 
+/**
+ The color applied that will be applied to message if there is a message. If there is no message
+ then this color will be applied to the title.
+ */
+@property(nonatomic, strong, nullable) UIColor *primaryTextColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Please add unit tests for all of these APIs.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 3]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
...
+ */
+@property(nonatomic, strong, nullable) UIColor *secondaryTextColor;
+/**
+ The color applied to the item list labels of Alert Controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *actionColor;
+/**
+ The color applied to the icon items of Alert Controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *imageColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Is this a tint color? If it is, let's keep `tint` in the name somewhere so clients know what it will do. Tint has a very specific application to images in UIKit.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 4]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
...
+ */
+@property(nonatomic, strong, nullable) UIColor *primaryTextColor;
+/**
+ The color applied to the title if there is both a title and a message.
+ */
+@property(nonatomic, strong, nullable) UIColor *secondaryTextColor;
+/**
+ The color applied to the item list labels of Alert Controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *actionColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Is this a text color? Maybe we can call it `actionTextColor`?

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 5]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
  */
 @property(nonatomic, nullable, strong) UIFont *actionFont;
 
+/**
+ The color applied to the sheet view of the action sheet controller.
+ */
 @property(nonatomic, nonnull, strong) UIColor *backgroundColor;
 
+/**
+ The color applied that will be applied to message if there is a message. If there is no message
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: This seems unnecessarily binding.  I think it's simpler if we provide simple properties mapping onto the view elements.  

If we have `titleTextColor` and `messageTextColor`, clients can choose how to apply the colors.  They will know if they have a title (and message) and can pick colors that way.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 6]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
...
+ then this color will be applied to the title.
+ */
+@property(nonatomic, strong, nullable) UIColor *primaryTextColor;
+/**
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Nit: Please add an extra line between the previous property and the start of comments.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 7]
```diff
@@ -140,8 +140,29 @@ __attribute__((objc_subclassing_restricted))
...
+ The color applied to the title if there is both a title and a message.
+ */
+@property(nonatomic, strong, nullable) UIColor *secondaryTextColor;
+/**
+ The color applied to the item list labels of Alert Controller.
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Shouldn't this be "Action Sheet" or something like that?

**Review by romoore — DISMISSED** [review]
Please add before/after screenshots.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 2 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: https://github.com/material-components/material-components-ios/pull/4993

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 3 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 4 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 5 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 6 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 7 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

**Comment by codeman7** [comment]
| Before | After |
| -------|------|
|simulator screen shot - iphone x - 2018-09-04 at 21 48 21|simulator screen shot - iphone x - 2018-09-04 at 21 47 10|

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m (hunk 1 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m]: I don't believe that will be needed, if it is then I will add it at a later date. This will have to be on the action items and if that is the case then tint color, actionTextColor, actionFont should all be on the action. As for now I believe maintaining this API I have proposed is the best solution.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 8]
```diff
@@ -150,18 +150,31 @@ __attribute__((objc_subclassing_restricted))
  then this color will be applied to the title.
  */
 @property(nonatomic, strong, nullable) UIColor *primaryTextColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: is `primaryTextColor` still required? It seems like it's better to have the new properties that assign values directly to the UI elements.

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 9]
```diff
@@ -188,15 +222,15 @@ typedef void (^MDCActionSheetHandler)(MDCActionSheetAction *_Nonnull action);
  Action sheet actions must have a title that will be set within actionWithTitle:image:handler:
  method.
  */
-@property (nonatomic, nonnull, readonly) NSString *title;
+@property(nonatomic, nonnull, readonly) NSString *title;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Since this is an `NSString`, please have it use `copy` semantics instead of `strong` (the default).

Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: It should also be marked `nullable` since I see the constructor being described in the docs actually accepts `nullable` parameters. 

>  \+ (nonnull instancetype)actionSheetControllerWithTitle:(nullable NSString *)title;

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 10]
```diff
@@ -114,6 +114,7 @@ - (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
     _header.title = [title copy];
     _header.message = [message copy];
     self.backgroundColor = [UIColor whiteColor];
+    self.actionTintColor = [UIColor blackColor];
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m]: Please don't use property setters in initializers. Directly access the autosynthesized ivar instead.  Please also use dot syntax for properties (on UIColor in this case).

```objc
_actionTintColor = UIColor.blackColor;
```

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 11]
```diff
@@ -114,6 +114,7 @@ - (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
     _header.title = [title copy];
     _header.message = [message copy];
     self.backgroundColor = [UIColor whiteColor];
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m]: Please don't call `self.backgroundColor` in the initializer. I see it's a custom setter and the object isn't finished being initialized yet.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 5 above)
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: Please remove `primaryTextColor` and `secondaryTextColor`

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h [hunk 12]
```diff
@@ -34,4 +34,12 @@
 
 @property (nonatomic, strong, nonnull) UIFont *messageFont;
 
+@property(nonatomic, strong, nullable) UIColor *primaryColor;
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: Please remove `primaryColor` and `secondaryColor`.

**Review by romoore — DISMISSED** [review]

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h (hunk 12 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: Primary color is required for a themer.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 8 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Primary text color is required for the themer.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 9 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 10 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m]: Done

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 11 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m]: Done

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h (hunk 12 above)
Comment (romoore) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: I believe the themer can simply set the titleTextColor and messageTextColor, correct? Why are separate properties required just for a themer?

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m (hunk 1 above)
Comment (romoore) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m]: Do you mean you will remove it at a later date? I'm proposing that the UIImage provided be the one directly applied and that we don't create a second image with a new rendering mode. This matches the behavior of UIKit and MDC components behave.

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.m [hunk 13]
```diff
@@ -201,4 +209,34 @@ - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
...
+
+- (void)setMessageTextColor:(UIColor *)messageTextColor {
+  _messageLabel.textColor = messageTextColor;
+}
+
+- (UIColor *)messageTextColor {
+  return _messageLabel.textColor;
+}
+
+- (BOOL)labelExist:(UILabel *)label {
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetHeaderView.m]: This method seems to be unused. Can it be removed?

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 2 above)
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: That PR has been closed. Are you planning to add the tests here to verify the behavior before merging?

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 14]
```diff
@@ -140,8 +140,42 @@ __attribute__((objc_subclassing_restricted))
...
+
+/**
+ The color applied to the item list labels of action sheet controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *actionTextColor;
+
+/**
+ The color applied to the icon items of action sheet controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *actionTintColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h:166]: Why not use `UIView.tintColor`? That would match what our other components do. Are we using that for some other part of the UI?

File: components/ActionSheet/src/MDCActionSheetController.h [hunk 15]
```diff
@@ -140,8 +140,42 @@ __attribute__((objc_subclassing_restricted))
...
+ */
+@property(nonatomic, strong, nullable) UIColor *secondaryTextColor;
+
+/**
+ The color applied to the title of the action sheet controller.
+ */
+@property(nonatomic, strong, nullable) UIColor *titleTextColor;
+
+/**
+ The color applied to the message of the actio sheet controller.
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Nit: s/actio/action/

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 16]
```diff
@@ -113,7 +114,8 @@ - (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
     _header = [[MDCActionSheetHeaderView alloc] initWithFrame:CGRectZero];
     _header.title = [title copy];
     _header.message = [message copy];
-    self.backgroundColor = [UIColor whiteColor];
+    _backgroundColor = UIColor.whiteColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m:117]: Do you need to set the `backgroundColor` values for `view`, `header`, and `tableView`?

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 17]
```diff
@@ -73,6 +73,7 @@ @implementation MDCActionSheetController {
 }
 
 @synthesize mdc_adjustsFontForContentSizeCategory = _mdc_adjustsFontForContentSizeCategory;
+@synthesize backgroundColor = _backgroundColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m:76]: This doesn't seem to be necessary. Previously you were using `self.view.backgroundColor` for the getter, and the setter assigned `backgroundColor` to `self.view`, `tableView`, and `backgroundColor`

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 18]
```diff
@@ -301,13 +305,46 @@ - (UIFont *)messageFont {
...
 - (void)setBackgroundColor:(UIColor *)backgroundColor {
+  _backgroundColor = backgroundColor;
   self.view.backgroundColor = backgroundColor;
   _tableView.backgroundColor = backgroundColor;
   _header.backgroundColor = backgroundColor;
 }
 
 - (UIColor *)backgroundColor {
-  return self.view.backgroundColor;
+  return _backgroundColor;
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m:317]: I think it makes sense to leave this as a pass-through to `self.view.backgroundColor`.

**Review by romoore — DISMISSED** [review]

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h (hunk 12 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: If there is only one label then that will be `onSurfaceColor` with `0.6` alpha. If there is both then the title will be `0.87` alpha. It depends if there is both labels or not and if a client applies a themer and then sets a label the theme won't be applied. `titleTextColor` and `messageTextColor` are not required, `primary` and `secondary` are.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 2 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: I will reopen that PR once this one lands.

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.m (hunk 13 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.m]: Done

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 15 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 17 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m:76]: Done

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 18 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m:317]: This has been removed.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 14 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:166]: The tint color is being applied to the image. A client may need an image that is red and an label that is blue.

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 16 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m:117]: Removed `self.view.backgroundColor`.

**Comment by codeman7** [comment]
Test can be tracked at https://github.com/material-components/material-components-ios/issues/5039

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 14 above)
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h:166]: I thought UILabel doesn't apply `tintColor` to the text color? At least I've never been able to get it to work.  Are you finding that `tintColor` is being applied to the label's text color in some circumstances?

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 16 above)
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.m:117]: I'm confused by this change. It looks like you're no longer setting the view's `backgroundColor` property, is that correct? Should these views initialize without a background color? Do `_header` and `_tableView` no longer require background colors either?

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 19]
```diff
@@ -301,13 +304,41 @@ - (UIFont *)messageFont {
 }
 
 - (void)setBackgroundColor:(UIColor *)backgroundColor {
-  self.view.backgroundColor = backgroundColor;
+  _backgroundColor = backgroundColor;
   _tableView.backgroundColor = backgroundColor;
   _header.backgroundColor = backgroundColor;
 }
 
-- (UIColor *)backgroundColor {
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m]: Why was this removed? It seems that for clients, the value of `- [controller backgroundColor]` will no longer be affected by `- [controller setBackgroundColor:]` or am I missing something?

**Review by romoore — DISMISSED** [review]
I'm most concerned about the conflicting behavior with `primaryTextColor` and `titleTextColor` and the apparent removal of support for `backgroundColor`. I would also appreciate replies to the unresolved comment threads from earlier reviews.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 5 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: I can remove `titleTextColor` and `messageTextColor` because those change based on each other. I cannot remove `primaryTextColor` and `secondaryTextColor` because those are needed for the color themer. I understand that you requested `titleTextColor` and `messageTextColor` but those changed based off of the `title` and `message` property. I believe that a better API is `primaryTextColor` and `secondaryTextColor`.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 9 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: I think you may be looking at the wrong initializer. The initializer for this `title` property is `(nonnull instancetype)actionWithTitle:(nonnull NSString *)title image:(nullable UIImage *)image handler:(__nullable MDCActionSheetHandler)handler;` you are referencing the initializer for `MDCActionSheetController`, this property is in `MDCActionSheetAction`.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 14 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:166]: If you are referring to the implementation why don't I apply `tintColor`, I think there is a disconnect because that is what is currently being used. If you are asking why the _API_ isn't `@property(nonatomic, strong, nullable) UIColor *tintColor` I feel that that can be confusing because clients may wonder what all this `tint` will be applied to. If you are referencing the `UIKit` equivalent `UIAlertController` with style `ActionSheet` doesn't have a `tintColor`. For now, I feel this is a better `API`.

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 16 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m:117]: I am now applying background color to the `view`, `_header`, and `_tableView`.

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 19 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m]: I added it back, if this is added then `@synthesize backgroundColor` is required, that is now added as well..

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m (hunk 1 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m]: If you don't set the `renderingMode` then the image will not be the correct color when you apply `tintColor` I can remove this line but in doing so it doesn't allow clients to set a `tintColor` on the images.

Comment (romoore) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m]: Clients can set the rendering mode. This matches what our other components do. The `tintColor` applied will be the one from the UIImageView, either inherited or explicitly applied.

**Comment by codeman7** [comment]
I have removed `primaryTextColor` and `secondaryTextColor`.

File: components/ActionSheet/src/MDCActionSheetController.h (hunk 14 above)
Comment (jverkoey) [reply components/ActionSheet/src/MDCActionSheetController.h:166]: There's some nuance here to ensure we consider:

1. If a client sets the tintColor of the action sheet, this will be propagated to any subviews that respond to tintColor. @codeman7 If we decide to go with this explicit actionTintColor API we need to make sure it plays nicely with the existing tintColor behavior. Have you explored that already?
2. I like the idea of having an explicit API that very clearly affects one thing, but 1. makes me a bit nervous that we might collide with expectations about how tintColor is propagating to our component.

**Comment by jverkoey** [comment]
This may be a good candidate PR to split apart along the lines of tint-based APIs and non-tint-based APIs because the questions around the tint-based APIs feel like they need a bit more exploration.
